[SPARK-36879][SQL] Support Parquet v2 data page encoding (DELTA_BINARY_PACKED) for the vectorized path

parthchandra opened the pull request:
### What changes were proposed in this pull request?
Implements a vectorized version of the parquet reader for DELTA_BINARY_PACKED encoding
This PR includes a previous PR for this issue which passed the read request thru to the parquet implementation and which was not vectorized. The current PR builds on top of that PR (hence both are included).

### Why are the changes needed?
Currently Spark throws an exception when reading data with these encodings if vectorized reader is enabled

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Additional unit tests for the encoding for both long and integer types (mirroring the unit tests in the Parquet implementation)

nkollar committed:
[SPARK-26509][SQL] Parquet DELTA_BYTE_ARRAY is not supported in Spark 2.x's Vectorized Reader

parthchandra committed:
[SPARK-36879][SQL] Vectorized implementation of DeltaBinaryPacked reader

SparkQA commented:
**[Test build #144861 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/144861/testReport)** for PR 34471 at commit [`fc9683b`](https://github.com/apache/spark/commit/fc9683bfc0af15050ddfdeef50a9ff7a45078bb8).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49330/

SparkQA commented:
**[Test build #144864 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/144864/testReport)** for PR 34471 at commit [`10659d3`](https://github.com/apache/spark/commit/10659d3bc9c1f3b3b9f8c29000f9f75e95281777).

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49330/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49330/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49334/

SparkQA commented:
**[Test build #144861 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/144861/testReport)** for PR 34471 at commit [`fc9683b`](https://github.com/apache/spark/commit/fc9683bfc0af15050ddfdeef50a9ff7a45078bb8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/144861/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49334/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49334/

SparkQA commented:
**[Test build #144864 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/144864/testReport)** for PR 34471 at commit [`10659d3`](https://github.com/apache/spark/commit/10659d3bc9c1f3b3b9f8c29000f9f75e95281777).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/144864/

dbtsai commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #144880 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/144880/testReport)** for PR 34471 at commit [`10659d3`](https://github.com/apache/spark/commit/10659d3bc9c1f3b3b9f8c29000f9f75e95281777).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49350/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49350/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49350/

SparkQA commented:
**[Test build #144880 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/144880/testReport)** for PR 34471 at commit [`10659d3`](https://github.com/apache/spark/commit/10659d3bc9c1f3b3b9f8c29000f9f75e95281777).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/144880/

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:
```diff
@@ -0,0 +1,319 @@
+        bitWidths[currentMiniBlock]);
+    for (int j = 0; j < miniBlockSizeInValues; j += 8) {
+      ByteBuffer buffer = in.slice(packer.getBitWidth());
+      packer.unpack8Values(buffer, buffer.position(), unpackedValuesBuffer, j);
```
wonder if we can unpack 32 values at once, since the [spec](https://github.com/apache/parquet-format/blob/master/Encodings.md#delta-encoding-delta_binary_packed--5) says the number of values in a mini block is a multiple of 32.

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:
```diff
@@ -0,0 +1,319 @@
+      long outValue = lastValueRead + minDeltaInCurrentBlock + unpackedValuesBuffer[i];
+      lastValueRead = outValue;
+      outputWriter.write(c, rowId + valuesRead, outValue);
+      remaining--;
```
hmm I wonder why we decrement `remaining` here.

sunchao commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetEncodingSuite.scala:
```diff
@@ -122,4 +126,40 @@ class ParquetEncodingSuite extends ParquetCompatibilityTest with SharedSparkSess
+    val hadoopConf = spark.sessionState.newHadoopConfWithOptions(extraOptions)
+    withSQLConf(
+      SQLConf.PARQUET_VECTORIZED_READER_ENABLED.key -> "true",
+      ParquetOutputFormat.JOB_SUMMARY_LEVEL -> "ALL") {
```
why we need this?

sunchao commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetDeltaEncodingIntegerSuite.scala:
```diff
@@ -0,0 +1,234 @@
+ * Translated from
+ *  org.apache.parquet.column.values.delta.DeltaBinaryPackingValuesWriterForIntegerTest
+ */
+class ParquetDeltaEncodingIntegerSuite extends ParquetCompatibilityTest with SharedSparkSession {
```
can we combine this with `ParquetDeltaEncodingLongSuite`? we can have a base class with the common code (need to abstract a few things like how to allocate an array, how to generate a rand int/long, and Parquet's value writer class etc.)

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaByteArrayReader.java:
```diff
@@ -0,0 +1,176 @@
+
+  @Override
+  public short readShort() {
+    return 0;
```
why return 0 here?

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaByteArrayReader.java:
```diff
@@ -0,0 +1,176 @@
+
+  @Override
+  public void readShorts(int total, WritableColumnVector c, int rowId) {
+
```
ditto

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:
```diff
@@ -0,0 +1,319 @@
+  @Override
+  public void readIntegersWithRebase(int total, WritableColumnVector c, int rowId,
+      boolean failIfRebase) {
+    throw new UnsupportedOperationException("Only readIntegers is valid.");
```
I think we'll need to implement these too.

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:
```diff
@@ -0,0 +1,319 @@
+  @Override
+  public void skipIntegers(int total) {
+    // Read the values but don't write them out (the writer output method is a no-op)
+    readValues(total, null, -1, (w, r, v) -> {
```
I think this can be done more efficiently, for instance we don't need to unpack the bits anymore, and don't need to compute the original value from delta, etc.

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaByteArrayReader.java:
```diff
@@ -0,0 +1,176 @@
+  }
+
+  @Override
+  public void skipBinary(int total) {
```
we need this too

sunchao reviewed (commented):
Thanks @parthchandra for working on this! I left some comments.

sunchao commented:
cc @sadikovi @viirya @dongjoon-hyun too

parthchandra commented:
Thank you for the review @sunchao! Let me address the comments.

dongjoon-hyun commented on .gitignore:
```diff
@@ -33,6 +33,7 @@ R/pkg/tests/fulltests/Rplots.pdf
 build/*.jar
 build/apache-maven*
 build/scala*
+build/zinc*
```
+1 with @sunchao 's comment. Please remove this from this PR.

dongjoon-hyun commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java:
```diff
@@ -280,6 +277,20 @@ private void initDataReader(
   }
 
+    private ValuesReader getValuesReader(Encoding encoding) {
+        switch (encoding) {
```
Could you use `two-space` indentation like the other part of this file, @parthchandra ?

dongjoon-hyun commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:
```diff
@@ -0,0 +1,319 @@
+import org.apache.parquet.io.api.Binary;
+import org.apache.spark.sql.execution.vectorized.WritableColumnVector;
+
+import java.io.IOException;
```
The import order is a little strange. Could you grouping `java` import (line 30 and 19) together as the first group?

dongjoon-hyun commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:
```diff
@@ -0,0 +1,319 @@
+
+    // read the first value
+    firstValue = BytesUtils.readZigZagVarLong(in);
+
```
nit. Let's remove redundant empty line.

dongjoon-hyun commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:
```diff
@@ -0,0 +1,319 @@
+  public void skipLongs(int total) {
+    // Read the values but don't write them out (the writer output method is a no-op)
+    readValues(total, null, -1, (w, r, v) -> {
+    });
```
Let's make these two lines into a single liner.
```scala
- readValues(total, null, -1, (w, r, v) -> {
- });
+ readValues(total, null, -1, (w, r, v) -> {});
```

dongjoon-hyun commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:
```diff
@@ -0,0 +1,319 @@
+    if (valuesRead + total > totalValueCount) {
+      throw new ParquetDecodingException(
+          "no more values to read, total value count is " + valuesRead);
+    }
```
Could you put this at the beginning before `int remaining = total;`?

parthchandra commented:
Looks like it may take some time to address some of the review comments. Marking this PR as draft in the meantime.

parthchandra committed:
Review comments addressed

SparkQA commented:
**[Test build #145121 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145121/testReport)** for PR 34471 at commit [`3572eaa`](https://github.com/apache/spark/commit/3572eaa42ce5360cc69b92888383bb39bef13a01).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49592/

SparkQA commented:
**[Test build #145122 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145122/testReport)** for PR 34471 at commit [`6021d90`](https://github.com/apache/spark/commit/6021d90a2a6bd9a90724bd30b4cffc999330f096).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49593/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49592/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49592/

SparkQA commented:
**[Test build #145125 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145125/testReport)** for PR 34471 at commit [`a88c721`](https://github.com/apache/spark/commit/a88c72169e7c3e624db1303ab5db1bfac253eb37).

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49593/

SparkQA commented:
**[Test build #145121 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145121/testReport)** for PR 34471 at commit [`3572eaa`](https://github.com/apache/spark/commit/3572eaa42ce5360cc69b92888383bb39bef13a01).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class VectorizedDeltaBinaryPackedReader extends VectorizedReaderBase `
  * `public class VectorizedDeltaByteArrayReader extends VectorizedReaderBase `
  * `public class VectorizedReaderBase extends ValuesReader implements VectorizedValuesReader `

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49593/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145121/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49596/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49596/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49596/

SparkQA commented:
**[Test build #145125 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145125/testReport)** for PR 34471 at commit [`a88c721`](https://github.com/apache/spark/commit/a88c72169e7c3e624db1303ab5db1bfac253eb37).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class VectorizedDeltaBinaryPackedReader extends VectorizedReaderBase `
  * `public class VectorizedDeltaByteArrayReader extends VectorizedReaderBase `
  * `public class VectorizedReaderBase extends ValuesReader implements VectorizedValuesReader `

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145125/

SparkQA commented:
**[Test build #145122 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145122/testReport)** for PR 34471 at commit [`6021d90`](https://github.com/apache/spark/commit/6021d90a2a6bd9a90724bd30b4cffc999330f096).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class VectorizedDeltaBinaryPackedReader extends VectorizedReaderBase `
  * `public class VectorizedDeltaByteArrayReader extends VectorizedReaderBase `
  * `public class VectorizedReaderBase extends ValuesReader implements VectorizedValuesReader `

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145122/

sunchao commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetTest.scala:
```diff
@@ -168,6 +169,16 @@ private[sql] trait ParquetTest extends FileBasedDataSourceTest {
   }
 
+  def withAllParquetWriters(code: => Unit): Unit = {
+    ParquetOutputFormat.WRITER_VERSION -> ParquetProperties.WriterVersion.PARQUET_2_0.toString
```
what is this line for?

sunchao reviewed (commented):
Looks much better now @parthchandra ! do you have any benchmark result to share, i.e., comparing the vectorized reader versus non-vectorized reader from `parquet-mr`? you can use `DataSourceReadBenchmark` for this.

parthchandra committed:
More review comments addressed

parthchandra committed:
Add benchmark.

parthchandra commented:
> Looks much better now @parthchandra ! do you have any benchmark result to share, i.e., comparing the vectorized reader versus non-vectorized reader from `parquet-mr`? you can use `DataSourceReadBenchmark` for this.

I've added a couple of cases to the benchmark for this. Because Delta encoding is more suited for sorted data, I also ran a modified version (not checked in) on sorted data. 

TL;DR For unsorted data Delta vectorized is slower than default (RLE) vectorized reader. For sorted data, Delta vectorized is much faster than unsorted data, but is still slightly slower than default (RLE) vectorized reader. For both unsorted/sorted data the vectorized reader is ~7x/~10x faster than the Parquet-MR reader.

Here is the summary for INT data

UNSORTED
```
OpenJDK 64-Bit Server VM 11.0.6+10-LTS on Mac OS X 10.15.7
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz

SQL Single INT Column Scan:               Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
SQL CSV                                           12625          12720         135          1.2         802.7       1.0X
SQL Json                                           6784           6844          85          2.3         431.3       1.9X
SQL Parquet Vectorized                              106            112           7        148.5           6.7     119.2X
SQL Parquet MR                                     1594           1601           9          9.9         101.4       7.9X
SQL ORC Vectorized                                  218            227           5         72.0          13.9      57.8X
SQL ORC MR                                         1289           1320          44         12.2          81.9       9.8X
SQL Parquet Vectorized (Delta Binary)               232            243          17         67.9          14.7      54.5X
SQL Parquet MR (Delta Binary)                      1462           1478          24         10.8          92.9       8.6X
```

SORTED 
```
OpenJDK 64-Bit Server VM 11.0.6+10-LTS on Mac OS X 10.15.7
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
SQL Single INT Column Scan:               Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
SQL CSV                                           12135          12150          22          1.3         771.5       1.0X
SQL Json                                           6138           6177          54          2.6         390.3       2.0X
SQL Parquet Vectorized                              102            107           7        153.5           6.5     118.4X
SQL Parquet MR                                     1489           1496           9         10.6          94.7       8.1X
SQL ORC Vectorized                                  166            171           3         94.9          10.5      73.2X
SQL ORC MR                                         1184           1188           5         13.3          75.3      10.3X
SQL Parquet Vectorized (Delta Binary)               130            138           8        120.7           8.3      93.1X
SQL Parquet MR (Delta Binary)                      1287           1306          26         12.2          81.8       9.4X
```

parthchandra commented:
I also checked on a real-world dataset. After writing the data compression was marginally better.
Parquet v1 (RLE) = 301.9 GB, Parquet v2 (DELTA_BINARY, DELTA_BYTE_ARRAY) = 299.9 GB. The bulk of the data was String data so this does not reflect the effectiveness of  DELTA_BINARY since most of the data would be encoded using DELTA_BYTE_ARRAY. 
Haven't been able to run a read benchmark on this yet.

SparkQA commented:
**[Test build #145367 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145367/testReport)** for PR 34471 at commit [`2574e69`](https://github.com/apache/spark/commit/2574e69577285075f102d7a39ef373ced8bdaaea).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49839/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/49839/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/49839/

SparkQA commented:
**[Test build #145367 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/145367/testReport)** for PR 34471 at commit [`2574e69`](https://github.com/apache/spark/commit/2574e69577285075f102d7a39ef373ced8bdaaea).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/145367/

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:
```diff
@@ -0,0 +1,315 @@
+    readValues(1, null, 0, (w, r, v) -> longVal = v);
+    return longVal;
+  }
+
```
nit: extra empty line

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:
```diff
@@ -0,0 +1,315 @@
+      IntegerOutputWriter outputWriter) {
+    if (valuesRead + total > totalValueCount) {
+      throw new ParquetDecodingException(
+          "no more values to read, total value count is " + valuesRead);
```
nit: let's all use capital letter for the first word to be consistent: "No more values to read ..."

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:
```diff
@@ -0,0 +1,315 @@
+  public final void readLongsWithRebase(
+      int total, WritableColumnVector c, int rowId, boolean failIfRebase) {
+    readValues(total, c, rowId, (w, r, v) -> {
+      if (v < RebaseDateTime.lastSwitchJulianDay()) {
```
should be `v < RebaseDateTime.lastSwitchJulianTs()`?

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:
```diff
@@ -0,0 +1,315 @@
+    try {
+      minDeltaInCurrentBlock = BytesUtils.readZigZagVarLong(in);
+    } catch (IOException e) {
+      throw new ParquetDecodingException("can not read min delta in current block", e);
```
ditto: "Can not read .."

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:
```diff
@@ -0,0 +1,315 @@
+      unpackMiniBlock();
+    }
+
+    //read values from miniblock
```
nit: space after `//`

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:
```diff
@@ -0,0 +1,315 @@
+    int valuesRead = 0;
+    for (int i = miniBlockSizeInValues - remainingInMiniBlock;
+        i < miniBlockSizeInValues && valuesRead < remaining; i++) {
+      //calculate values from deltas unpacked for current block
```
ditto

sunchao commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:
```diff
@@ -0,0 +1,315 @@
+   * see org.apache.parquet.column.values.delta.DeltaBinaryPackingValuesReader#unpackMiniBlock
+   */
+  private void unpackMiniBlock() throws IOException {
+    Arrays.fill(this.unpackedValuesBuffer, 0);
```
Is this necessary? since we are overwriting it anyway.

sunchao commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetEncodingSuite.scala:
```diff
@@ -16,14 +16,22 @@
+import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.test.SharedSparkSession
 
+
```
nit: unnecessary change.

sunchao commented on sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DataSourceReadBenchmark.scala:
```diff
@@ -147,6 +159,16 @@ object DataSourceReadBenchmark extends SqlBasedBenchmark {
           }
         }
 
+        sqlBenchmark.addCase("SQL Parquet Vectorized (Delta Binary)") { _ =>
```
Can we:
- move this up to be adjacent to the Parquet cases
- change the name to "SQL Parquet Vectorized Format v2" or something? 

You also need to rebase and handle the newly added `BooleanType` here.

parthchandra replied:
I guess I didn't understand the implication when you suggested I need to rebase!
It looks like the Boolean type breaks vectorized reading. The boolean column gets encoded as RLE but the vectorized RLE reader does not support Boolean.
```
[info] Caused by: java.lang.UnsupportedOperationException: only readInts is valid.
[info] 	at org.apache.spark.sql.execution.datasources.parquet.VectorizedRleValuesReader.readBooleans(VectorizedRleValuesReader.java:368)
```
I guess I could fix this. Would you recommend a different PR, or should I make the change in this one?

sunchao replied:
Right, Boolean is encoded as RLE in Parquet format V2, so we'd need to handle it too. I think we should raise a separate PR for this.

For now, can you remove the changes in `DataSourceReadBenchmark` and split it out into a separate PR later once we have the boolean type supported?

sunchao reviewed (commented):
Thanks @parthchandra for the update and sorry for my late response!

The benchmark looks good. I think it's definitely worth doing because otherwise users will get exception instead and have to resort to the much slower `parquet-mr` implementation.

parthchandra committed:
more review comments addressed

parthchandra reviewed (commented):
Addressed review comments and rebased

SparkQA commented:
**[Test build #146144 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146144/testReport)** for PR 34471 at commit [`91f7534`](https://github.com/apache/spark/commit/91f7534c87c67e6e456a06f6f250dd63e8b14f3c).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50617/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50617/

SparkQA commented:
**[Test build #146144 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146144/testReport)** for PR 34471 at commit [`91f7534`](https://github.com/apache/spark/commit/91f7534c87c67e6e456a06f6f250dd63e8b14f3c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50617/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146144/

sunchao reviewed (commented):
Thanks a lot for updating this @parthchandra ! Overall look pretty good. I think we just need to address the issue with benchmark and attach the result together with the PR. You can find out how to get benchmark result using GitHub workflow [here](https://spark.apache.org/developer-tools.html).

parthchandra committed:
More review comments addressed. Removed benchmark (to be added back once Booleans/RLE is implemented)

SparkQA commented:
**[Test build #146200 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146200/testReport)** for PR 34471 at commit [`f472135`](https://github.com/apache/spark/commit/f472135a04a3129914b94ddddc03d506d07626a1).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50674/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50674/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50674/

SparkQA commented:
**[Test build #146200 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146200/testReport)** for PR 34471 at commit [`f472135`](https://github.com/apache/spark/commit/f472135a04a3129914b94ddddc03d506d07626a1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146200/

sunchao commented:
@parthchandra could you address the unit tests failure?

parthchandra commented:
The unit tests are failing in parts that I am not familiar with. Previously, re-running the tests had worked, but this time around the tests are failing every time. Can I get some help figuring out where the problem is?

sunchao commented:
If you click the SparkQA test build link you should see the failed tests. For instance:

```
sbt.ForkMain$ForkError: org.scalatest.exceptions.TestFailedException: "[no more values to read, total value count is 641]" did not equal "[No more values to read. Total values read:  641, total count: 641, trying to read 1 more.]"
	at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:472)
	at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:471)
	at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1231)
	at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:1295)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetDeltaEncodingSuite.$anonfun$new$21(ParquetDeltaEncodingSuite.scala:170)
	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
```

parthchandra committed:
Fix broken unit test

parthchandra commented:
Thank you @sunchao. I had gone thru the log and failed to see the test(s) that had failed. One of the unit tests was checking for the error message that accompanied an exception and as part of the review I had changed the error message! Updated the test. The tests should pass now.

SparkQA commented:
**[Test build #146491 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146491/testReport)** for PR 34471 at commit [`64cc82f`](https://github.com/apache/spark/commit/64cc82fc2898cf553900b589f93bdd4a5887060d).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50967/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50967/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50967/

SparkQA commented:
**[Test build #146491 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146491/testReport)** for PR 34471 at commit [`64cc82f`](https://github.com/apache/spark/commit/64cc82fc2898cf553900b589f93bdd4a5887060d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146491/

sunchao reviewed (approved):
LGTM

parthchandra committed:
fix rebase conflicts

sunchao commented:
Committed to master branch, thanks @parthchandra !

LuciferYang commented on sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DataSourceReadBenchmark.scala:
```diff
@@ -76,6 +79,7 @@ object DataSourceReadBenchmark extends SqlBasedBenchmark {
     saveAsCsvTable(testDf, dir.getCanonicalPath + "/csv")
     saveAsJsonTable(testDf, dir.getCanonicalPath + "/json")
     saveAsParquetTable(testDf, dir.getCanonicalPath + "/parquet")
+    saveAsParquetV2Table(testDf, dir.getCanonicalPath + "/parquetV2")
```
Maybe we should update the benchmark-result of `DataSourceReadBenchmark` @parthchandra parthchandra

LuciferYang replied:
I found that there are still unsupported encoding in Data Page V2, such as RLE. It seems that it is not time to update the benchmark, please ignore my previous comments

parthchandra replied:
@LuciferYang I was getting ready to set up a PR for the RLE/Boolean encoding and noticed that you have done so. Thank you!
Adding back the benchmark in a new PR.

LuciferYang commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaBinaryPackedReader.java:
```diff
@@ -0,0 +1,319 @@
+  private ByteBufferInputStream in;
+
+  // temporary buffers used by readByte, readShort, readInteger, and readLong
+  byte byteVal;
```
Should these 4 field be private?

MaxGekk commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRebaseDatetimeSuite.scala:
```diff
@@ -134,33 +134,35 @@ abstract class ParquetRebaseDatetimeSuite
-        withSQLConf(SQLConf.PARQUET_OUTPUT_TIMESTAMP_TYPE.key -> tsOutputType) {
+          val df = Seq.tabulate(N)(rowFunc).toDF("dict", "plain")
+            .select($"dict".cast(catalystType), $"plain".cast(catalystType))
+          withSQLConf(SQLConf.PARQUET_OUTPUT_TIMESTAMP_TYPE.key -> tsOutputType) {
```
Something wrong with indentation here and below:
<img width="642" alt="Screenshot 2022-01-06 at 22 56 02" src="https://user-images.githubusercontent.com/1580697/148443305-f6890111-26c2-4538-9abe-53a28144c46b.png">

MaxGekk commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRebaseDatetimeSuite.scala:
```diff
@@ -219,60 +221,62 @@ abstract class ParquetRebaseDatetimeSuite
                         Seq.tabulate(N)(_ => Row(nonRebased)))
                     }
                   }
+                  }
```
Wrong indentation too
